File: memory/allocator.c
/*
 * allocator.c - TLSF-Inspired Memory Allocator Implementation
 * ============================================================
 *
 * See allocator.h for design overview and API documentation.
 *
 * IMPLEMENTATION DETAILS:
 * -----------------------
 *
 * BLOCK HEADER STRUCTURE:
 * The size_and_flags field packs the block size and status flags:
 *   Bits [63:2] = Size (always multiple of 8)
 *   Bit 1 = PREV_FREE (previous block in memory is free)
 *   Bit 0 = FREE (this block is free)
 *
 * FREE BLOCK LAYOUT:
 *   +------------------+
 *   | size_and_flags   |  Header (8 bytes)
 *   +------------------+
 *   | next_free        |  Pointer to next in free list (8 bytes)
 *   +------------------+
 *   | prev_free        |  Pointer to prev in free list (8 bytes)
 *   +------------------+
 *   |    ...           |  (unused space)
 *   +------------------+
 *   | footer (size)    |  Copy of size for backwards traversal (8 bytes)
 *   +------------------+
 *
 * ALLOCATED BLOCK LAYOUT:
 *   +------------------+
 *   | size_and_flags   |  Header (8 bytes)
 *   +------------------+
 *   |                  |
 *   |   User Data      |  (payload area)
 *   |                  |
 *   +------------------+
 *   (No footer needed - we only need footer for coalescing with free blocks)
 *
 * SIZE CLASS CALCULATION:
 * Given a block size, we compute its size class as:
 *   class = floor(log2(size)) - floor(log2(MIN_BLOCK_SIZE))
 *
 * This groups blocks into power-of-two buckets:
 *   Class 0: [32, 64)
 *   Class 1: [64, 128)
 *   Class 2: [128, 256)
 *   etc.
 */

#include "allocator.h"
#include "../common/string.h"

/* =============================================================================
 * INTERNAL DATA STRUCTURES
 * =============================================================================
 */

/*
 * Block header
 *
 * Present at the start of every block (allocated or free).
 */
typedef struct block_header {
    uintptr_t size_and_flags;
} block_header_t;

/*
 * Free list node
 *
 * Stored in the payload area of free blocks.
 * Forms a doubly-linked list for each size class.
 */
typedef struct free_node {
    struct free_node *next;
    struct free_node *prev;
} free_node_t;


/* =============================================================================
 * HELPER FUNCTIONS - ALIGNMENT
 * =============================================================================
 */

/*
 * align_up() - Round up to alignment boundary
 */
static inline uintptr_t align_up(uintptr_t addr, size_t align)
{
    return (addr + align - 1) & ~(align - 1);
}


/* =============================================================================
 * HELPER FUNCTIONS - BLOCK HEADER OPERATIONS
 * =============================================================================
 */

/*
 * block_size() - Get block size (excluding flags)
 */
static inline size_t block_size(const block_header_t *block)
{
    return block->size_and_flags & ~(uintptr_t)FLAG_MASK;
}

/*
 * block_set_size() - Set block size (preserving flags)
 */
static inline void block_set_size(block_header_t *block, size_t size)
{
    block->size_and_flags = size | (block->size_and_flags & FLAG_MASK);
}

/*
 * block_is_free() - Check if block is free
 */
static inline bool block_is_free(const block_header_t *block)
{
    return (block->size_and_flags & FLAG_FREE) != 0;
}

/*
 * block_set_free() - Mark block as free or allocated
 */
static inline void block_set_free(block_header_t *block, bool free)
{
    if (free) {
        block->size_and_flags |= FLAG_FREE;
    } else {
        block->size_and_flags &= ~(uintptr_t)FLAG_FREE;
    }
}

/*
 * block_is_prev_free() - Check if previous block is free
 */
static inline bool block_is_prev_free(const block_header_t *block)
{
    return (block->size_and_flags & FLAG_PREV_FREE) != 0;
}

/*
 * block_set_prev_free() - Mark whether previous block is free
 */
static inline void block_set_prev_free(block_header_t *block, bool prev_free)
{
    if (prev_free) {
        block->size_and_flags |= FLAG_PREV_FREE;
    } else {
        block->size_and_flags &= ~(uintptr_t)FLAG_PREV_FREE;
    }
}

/*
 * block_payload() - Get pointer to payload area
 */
static inline void *block_payload(block_header_t *block)
{
    return (uint8_t *)block + HEADER_SIZE;
}

/*
 * block_from_payload() - Get header from payload pointer
 */
static inline block_header_t *block_from_payload(void *payload)
{
    return (block_header_t *)((uint8_t *)payload - HEADER_SIZE);
}

/*
 * block_next() - Get next block in memory
 */
static inline block_header_t *block_next(block_header_t *block)
{
    return (block_header_t *)((uint8_t *)block + block_size(block));
}

/*
 * block_prev() - Get previous block (only valid if prev_free flag set)
 *
 * When a block is free, it has a footer (copy of size) at its end.
 * We can use this to find where the previous block started.
 */
static inline block_header_t *block_prev(block_header_t *block)
{
    /*
     * Footer is at (block - 8 bytes), contains size of previous block.
     * Previous block starts at (block - prev_size).
     */
    uintptr_t *footer = (uintptr_t *)block - 1;
    size_t prev_size = *footer;
    return (block_header_t *)((uint8_t *)block - prev_size);
}


/* =============================================================================
 * HELPER FUNCTIONS - FREE BLOCK OPERATIONS
 * =============================================================================
 */

/*
 * free_list_node() - Get free list node from block header
 */
static inline free_node_t *free_list_node(block_header_t *block)
{
    return (free_node_t *)block_payload(block);
}

/*
 * block_from_node() - Get block header from free list node
 */
static inline block_header_t *block_from_node(free_node_t *node)
{
    return (block_header_t *)((uint8_t *)node - HEADER_SIZE);
}

/*
 * write_footer() - Write size to block's footer
 *
 * The footer is used to find the previous block during coalescing.
 */
static inline void write_footer(block_header_t *block)
{
    size_t size = block_size(block);
    uintptr_t *footer = (uintptr_t *)((uint8_t *)block + size - FOOTER_SIZE);
    *footer = size;
}

/*
 * size_class() - Calculate size class for a block size
 *
 * Uses the position of the highest set bit to determine class.
 * Equivalent to floor(log2(size)) - floor(log2(MIN_BLOCK_SIZE))
 */
static inline size_t size_class(size_t size)
{
    if (size <= MIN_BLOCK_SIZE) {
        return 0;
    }

    /*
     * Count leading zeros to find highest bit position.
     * class = (bit_position(size) - bit_position(MIN_BLOCK_SIZE))
     *
     * MIN_BLOCK_SIZE = 32 = 2^5, so its highest bit is at position 5.
     * For size 64 (2^6), highest bit at 6, class = 6-5 = 1.
     * For size 128 (2^7), highest bit at 7, class = 7-5 = 2.
     */
    unsigned int bits = 0;
    size_t temp = size;
    while (temp > 1) {
        temp >>= 1;
        bits++;
    }

    unsigned int min_bits = 5;  /* log2(32) = 5 */
    size_t class = (bits > min_bits) ? (bits - min_bits) : 0;

    /* Clamp to valid range */
    if (class >= NUM_SIZE_CLASSES) {
        class = NUM_SIZE_CLASSES - 1;
    }

    return class;
}


/* =============================================================================
 * ALLOCATOR OPERATIONS - FREE LIST MANAGEMENT
 * =============================================================================
 */

/*
 * remove_from_free_list() - Remove a block from its free list
 */
static void remove_from_free_list(allocator_t *alloc, block_header_t *block)
{
    size_t size = block_size(block);
    size_t class = size_class(size);
    free_node_t *node = free_list_node(block);

    /* Update previous node's next pointer (or list head) */
    if (node->prev) {
        node->prev->next = node->next;
    } else {
        /* This was the head of the list */
        alloc->free_lists[class] = node->next;
    }

    /* Update next node's prev pointer */
    if (node->next) {
        node->next->prev = node->prev;
    }

    /* If list is now empty, clear bitmap bit */
    if (alloc->free_lists[class] == NULL) {
        alloc->free_bitmap &= ~(1u << class);
    }
}

/*
 * insert_into_free_list() - Add a block to its free list
 */
static void insert_into_free_list(allocator_t *alloc, block_header_t *block)
{
    size_t size = block_size(block);
    size_t class = size_class(size);
    free_node_t *node = free_list_node(block);

    /* Insert at head of list */
    free_node_t *old_head = (free_node_t *)alloc->free_lists[class];
    node->prev = NULL;
    node->next = old_head;

    if (old_head) {
        old_head->prev = node;
    }

    alloc->free_lists[class] = node;

    /* Set bitmap bit */
    alloc->free_bitmap |= (1u << class);
}


/* =============================================================================
 * ALLOCATOR OPERATIONS - BLOCK SPLITTING AND COALESCING
 * =============================================================================
 */

/*
 * split_block() - Split a block if it's larger than needed
 *
 * If the remainder after allocation is large enough to be a valid block,
 * create a new free block with the leftover space.
 */
static void split_block(allocator_t *alloc, block_header_t *block, size_t needed)
{
    size_t current_size = block_size(block);
    size_t remainder = current_size - needed;

    /* Only split if remainder can form a valid block */
    if (remainder < MIN_BLOCK_SIZE) {
        return;
    }

    /* Shrink original block */
    block_set_size(block, needed);

    /* Create remainder block */
    block_header_t *rem = (block_header_t *)((uint8_t *)block + needed);
    rem->size_and_flags = remainder | FLAG_FREE;
    block_set_prev_free(rem, false);

    /* Initialize free list node for remainder */
    free_node_t *rem_node = free_list_node(rem);
    rem_node->next = NULL;
    rem_node->prev = NULL;

    /* Write footer for remainder block */
    write_footer(rem);

    /* Add remainder to free list */
    insert_into_free_list(alloc, rem);

    /* Update next block's prev_free flag */
    uintptr_t next_addr = (uintptr_t)block_next(rem);
    if (next_addr < alloc->heap_end) {
        block_header_t *next = (block_header_t *)next_addr;
        block_set_prev_free(next, true);
    }
}

/*
 * coalesce() - Merge adjacent free blocks
 *
 * After freeing a block, check if neighbors are also free and merge them.
 * This prevents fragmentation from accumulating over time.
 */
static block_header_t *coalesce(allocator_t *alloc, block_header_t *block)
{
    /*
     * Check next block
     *
     * If the next block in memory is free, absorb it into this block.
     */
    uintptr_t next_addr = (uintptr_t)block_next(block);
    if (next_addr < alloc->heap_end) {
        block_header_t *next = (block_header_t *)next_addr;
        if (block_is_free(next)) {
            /* Remove next from free list */
            remove_from_free_list(alloc, next);

            /* Absorb next into current block */
            size_t new_size = block_size(block) + block_size(next);
            block_set_size(block, new_size);
        }
    }

    /*
     * Check previous block
     *
     * If the previous block is free (indicated by our prev_free flag),
     * we should merge into it instead.
     */
    if (block_is_prev_free(block)) {
        block_header_t *prev = block_prev(block);

        /* Remove prev from free list */
        remove_from_free_list(alloc, prev);

        /* Absorb current block into prev */
        size_t new_size = block_size(prev) + block_size(block);
        block_set_size(prev, new_size);

        /* Continue with prev as the block to insert */
        block = prev;
    }

    /* Write footer for the (possibly merged) block */
    write_footer(block);

    /* Update next block's prev_free flag */
    next_addr = (uintptr_t)block_next(block);
    if (next_addr < alloc->heap_end) {
        block_header_t *next = (block_header_t *)next_addr;
        block_set_prev_free(next, true);
    }

    return block;
}

/*
 * find_free_block() - Find a free block of at least the given size
 *
 * Uses the bitmap to quickly find a size class with available blocks.
 */
static block_header_t *find_free_block(allocator_t *alloc, size_t size)
{
    size_t min_class = size_class(size);

    /*
     * Mask off classes smaller than we need, then find first set bit.
     * This gives us the smallest class that might have a large enough block.
     */
    uint32_t available = alloc->free_bitmap & ~((1u << min_class) - 1);

    if (available == 0) {
        return NULL;  /* No suitable blocks */
    }

    /*
     * Find first set bit (lowest class with blocks)
     */
    size_t class = 0;
    while ((available & (1u << class)) == 0) {
        class++;
    }

    /* Get first block from that class */
    free_node_t *node = (free_node_t *)alloc->free_lists[class];
    if (node == NULL) {
        return NULL;  /* Shouldn't happen if bitmap is correct */
    }

    return block_from_node(node);
}


/* =============================================================================
 * PUBLIC API IMPLEMENTATION
 * =============================================================================
 */

/*
 * allocator_init() - Initialize allocator with memory region
 */
void allocator_init(allocator_t *alloc, uintptr_t heap_start, uintptr_t heap_end)
{
    /* Align heap bounds */
    uintptr_t start = align_up(heap_start, MIN_ALIGN);
    uintptr_t end = heap_end & ~(uintptr_t)(MIN_ALIGN - 1);
    size_t size = end - start;

    /* Initialize allocator state */
    alloc->heap_start = start;
    alloc->heap_end = end;
    alloc->free_space = size;
    alloc->allocated = 0;
    alloc->free_bitmap = 0;

    /* Clear free lists */
    for (int i = 0; i < NUM_SIZE_CLASSES; i++) {
        alloc->free_lists[i] = NULL;
    }

    /*
     * Create one large free block spanning the entire heap
     */
    block_header_t *initial = (block_header_t *)start;
    initial->size_and_flags = size | FLAG_FREE;

    /* Initialize free list node */
    free_node_t *node = free_list_node(initial);
    node->next = NULL;
    node->prev = NULL;

    /* Write footer */
    write_footer(initial);

    /* Add to free list */
    size_t class = size_class(size);
    alloc->free_lists[class] = node;
    alloc->free_bitmap = 1u << class;

    alloc->initialized = true;
}

/*
 * allocator_init_from_ram() - Initialize from RAM detection results
 */
void allocator_init_from_ram(allocator_t *alloc,
                             uintptr_t ram_base, size_t ram_size,
                             uintptr_t heap_symbol,
                             uintptr_t *heap_start_out, uintptr_t *heap_end_out,
                             uintptr_t *jit_start_out, uintptr_t *jit_end_out)
{
    uintptr_t ram_end = ram_base + ram_size;

    /* JIT region at end of RAM */
    uintptr_t jit_start = ram_end - JIT_SIZE;
    uintptr_t jit_end = ram_end;

    /* Heap ends where JIT begins */
    uintptr_t heap_start = heap_symbol;
    uintptr_t heap_end = jit_start;

    /* Initialize allocator */
    allocator_init(alloc, heap_start, heap_end);

    /* Return bounds */
    if (heap_start_out) *heap_start_out = heap_start;
    if (heap_end_out) *heap_end_out = heap_end;
    if (jit_start_out) *jit_start_out = jit_start;
    if (jit_end_out) *jit_end_out = jit_end;
}

/*
 * allocator_alloc() - Allocate memory
 */
void *allocator_alloc(allocator_t *alloc, size_t size, size_t align)
{
    if (!alloc->initialized) {
        return NULL;
    }

    if (size == 0) {
        return NULL;
    }

    /* Ensure minimum alignment */
    if (align < MIN_ALIGN) {
        align = MIN_ALIGN;
    }

    /*
     * Calculate total block size needed
     *
     * Payload must be large enough for free list node when freed.
     */
    size_t payload_size = size;
    if (payload_size < sizeof(free_node_t)) {
        payload_size = sizeof(free_node_t);
    }

    size_t block_size_needed = align_up(HEADER_SIZE + payload_size, MIN_ALIGN);
    if (block_size_needed < MIN_BLOCK_SIZE) {
        block_size_needed = MIN_BLOCK_SIZE;
    }

    /* Find a suitable free block */
    block_header_t *block = find_free_block(alloc, block_size_needed);
    if (block == NULL) {
        return NULL;  /* Out of memory */
    }

    /* Remove from free list */
    remove_from_free_list(alloc, block);

    /* Split if block is much larger than needed */
    split_block(alloc, block, block_size_needed);

    /* Mark as allocated */
    block_set_free(block, false);

    /* Update next block's prev_free flag */
    size_t final_size = block_size(block);
    uintptr_t next_addr = (uintptr_t)block_next(block);
    if (next_addr < alloc->heap_end) {
        block_header_t *next = (block_header_t *)next_addr;
        block_set_prev_free(next, false);
    }

    /* Update statistics */
    alloc->allocated += final_size;
    alloc->free_space -= final_size;

    return block_payload(block);
}

/*
 * allocator_free() - Free previously allocated memory
 */
void allocator_free(allocator_t *alloc, void *ptr)
{
    if (ptr == NULL) {
        return;
    }

    if (!alloc->initialized) {
        return;
    }

    /* Get block header */
    block_header_t *block = block_from_payload(ptr);
    size_t size = block_size(block);

    /* Update statistics */
    alloc->allocated -= size;
    alloc->free_space += size;

    /* Mark as free */
    block_set_free(block, true);

    /* Coalesce with neighbors */
    block = coalesce(alloc, block);

    /* Insert into free list */
    insert_into_free_list(alloc, block);
}

/*
 * allocator_realloc() - Reallocate memory
 */
void *allocator_realloc(allocator_t *alloc, void *ptr, size_t new_size)
{
    if (ptr == NULL) {
        return allocator_alloc(alloc, new_size, MIN_ALIGN);
    }

    if (new_size == 0) {
        allocator_free(alloc, ptr);
        return NULL;
    }

    /* Get current block info */
    block_header_t *block = block_from_payload(ptr);
    size_t current_size = block_size(block) - HEADER_SIZE;

    /* If shrinking or same size, just return current pointer */
    if (new_size <= current_size) {
        /* Could potentially split here, but keeping it simple */
        return ptr;
    }

    /* Need to grow - allocate new block and copy */
    void *new_ptr = allocator_alloc(alloc, new_size, MIN_ALIGN);
    if (new_ptr == NULL) {
        return NULL;  /* Original pointer still valid */
    }

    /* Copy old data */
    memcpy(new_ptr, ptr, current_size);

    /* Free old block */
    allocator_free(alloc, ptr);

    return new_ptr;
}

/*
 * allocator_stats() - Get statistics
 */
void allocator_stats(const allocator_t *alloc, size_t *allocated_out, size_t *free_out)
{
    if (allocated_out) *allocated_out = alloc->allocated;
    if (free_out) *free_out = alloc->free_space;
}

/*
 * allocator_bounds() - Get heap bounds
 */
void allocator_bounds(const allocator_t *alloc, uintptr_t *start_out, uintptr_t *end_out)
{
    if (start_out) *start_out = alloc->heap_start;
    if (end_out) *end_out = alloc->heap_end;
}

/*
 * allocator_is_initialized() - Check initialization
 */
bool allocator_is_initialized(const allocator_t *alloc)
{
    return alloc->initialized;
}


/* =============================================================================
 * GLOBAL ALLOCATOR
 * =============================================================================
 */

/* Global allocator instance */
allocator_t g_allocator = {0};

/* Linker symbol for heap start - must be defined in linker script */
extern uint8_t __heap_start[];

void heap_init(uintptr_t ram_base, size_t ram_size)
{
    uintptr_t heap_start, heap_end, jit_start, jit_end;
    allocator_init_from_ram(&g_allocator,
                            ram_base, ram_size,
                            (uintptr_t)__heap_start,
                            &heap_start, &heap_end,
                            &jit_start, &jit_end);
}

void *heap_alloc(size_t size)
{
    return allocator_alloc(&g_allocator, size, MIN_ALIGN);
}

void *heap_alloc_aligned(size_t size, size_t align)
{
    return allocator_alloc(&g_allocator, size, align);
}

void heap_free(void *ptr)
{
    allocator_free(&g_allocator, ptr);
}

void *heap_realloc(void *ptr, size_t size)
{
    return allocator_realloc(&g_allocator, ptr, size);
}

File: memory/src/allocator.rs
//! TLSF-inspired memory allocator for bare-metal Tutorial-OS.
//!
//! This allocator provides `O(1)` allocation and deallocation with bounded
//! fragmentation, suitable for a bare-metal OS that cannot fall back to
//! an operating system allocator.
//!
//! # Design
//!
//! TLSF (Two-Level Segregated Fit) organises free blocks into a two-level
//! bitmap index.  The first level divides sizes by powers of two; the
//! second level subdivides each power-of-two range into linear segments.
//! Finding a suitable free block requires only two bitmap scans (each a
//! single `CTZ` or `CLZ` instruction on all supported architectures).
//!
//! # Usage
//!
//! ```ignore
//! static mut HEAP: [u8; 1024 * 1024] = [0; 1024 * 1024];
//!
//! let mut alloc = Allocator::new();
//! unsafe { alloc.init(HEAP.as_mut_ptr(), HEAP.len()); }
//!
//! let ptr = alloc.alloc(256, 8);  // 256 bytes, 8-byte aligned
//! alloc.free(ptr);
//! ```
//!
//! # Safety
//!
//! This is inherently unsafe — it manages raw memory regions.  The caller
//! must ensure the heap region is valid, exclusively owned, and not
//! accessed through any other means while the allocator is in use.

use core::ptr;

// ============================================================================
// Configuration
// ============================================================================

/// Number of first-level index classes (each is a power-of-two range).
const FL_INDEX_COUNT: usize = 20; // Covers up to 2^(FL_INDEX_SHIFT + 20 - 1) ≈ 512 MB

/// Number of second-level subdivisions within each first-level class.
const SL_INDEX_COUNT_LOG2: usize = 4;
const SL_INDEX_COUNT: usize = 1 << SL_INDEX_COUNT_LOG2; // 16

/// Minimum allocation size (must be large enough for the free block header).
const MIN_BLOCK_SIZE: usize = 32;

/// First-level index shift — the smallest first-level class covers
/// blocks of size `[2^FL_INDEX_SHIFT, 2^(FL_INDEX_SHIFT+1))`.
const FL_INDEX_SHIFT: usize = 5; // 32 bytes

// ============================================================================
// Block Header
// ============================================================================

/// Bit flag: block is free.
const BLOCK_FREE: usize = 1 << 0;
/// Bit flag: the preceding physical block is free.
const BLOCK_PREV_FREE: usize = 1 << 1;
/// Mask for the size field (excludes flag bits).
const SIZE_MASK: usize = !(BLOCK_FREE | BLOCK_PREV_FREE);

/// Header placed at the start of every memory block (allocated or free).
///
/// Free blocks additionally contain `next_free` / `prev_free` pointers
/// that overlay the user data region.
#[repr(C)]
struct BlockHeader {
    /// Encodes the block size (in bytes, including the header) plus flags
    /// in the low two bits.
    size_and_flags: usize,
    /// Pointer to the preceding physical block (for coalescing on free).
    prev_phys: *mut BlockHeader,
    // --- The following fields exist only in FREE blocks ---
    // (They overlap the user data area of the block.)
    next_free: *mut BlockHeader,
    prev_free: *mut BlockHeader,
}

#[allow(dead_code)]
impl BlockHeader {
    #[inline]
    fn size(&self) -> usize {
        self.size_and_flags & SIZE_MASK
    }

    #[inline]
    fn set_size(&mut self, size: usize) {
        self.size_and_flags = size | (self.size_and_flags & !SIZE_MASK);
    }

    #[inline]
    fn is_free(&self) -> bool {
        self.size_and_flags & BLOCK_FREE != 0
    }

    #[inline]
    fn set_free(&mut self) {
        self.size_and_flags |= BLOCK_FREE;
    }

    #[inline]
    fn set_used(&mut self) {
        self.size_and_flags &= !BLOCK_FREE;
    }

    #[inline]
    fn is_prev_free(&self) -> bool {
        self.size_and_flags & BLOCK_PREV_FREE != 0
    }

    #[inline]
    fn set_prev_free(&mut self) {
        self.size_and_flags |= BLOCK_PREV_FREE;
    }

    #[inline]
    fn clear_prev_free(&mut self) {
        self.size_and_flags &= !BLOCK_PREV_FREE;
    }

    /// User data starts immediately after `size_and_flags` and `prev_phys`.
    #[inline]
    fn payload(&self) -> *mut u8 {
        let hdr_ptr = self as *const Self as *mut u8;
        unsafe { hdr_ptr.add(core::mem::size_of::<usize>() * 2) }
    }

    /// Get the next physical block (adjacent in memory).
    #[inline]
    unsafe fn next_phys(&self) -> *mut BlockHeader {
        let base = self as *const Self as *const u8;
        unsafe { base.add(self.size()) as *mut BlockHeader }
    }
}

// ============================================================================
// Allocator
// ============================================================================

/// TLSF-inspired memory allocator.
pub struct Allocator {
    /// First-level bitmap — bit `i` is set if any SL list in FL class `i` is non-empty.
    fl_bitmap: u32,
    /// Second-level bitmaps — one per FL class.
    sl_bitmaps: [u16; FL_INDEX_COUNT],
    /// Free list heads: `blocks[fl][sl]` points to the first free block in
    /// that size class, or null.
    blocks: [[*mut BlockHeader; SL_INDEX_COUNT]; FL_INDEX_COUNT],
    /// Total bytes managed.
    heap_size: usize,
    /// Total bytes currently allocated (including headers, for diagnostics).
    used: usize,
}

// SAFETY: bare-metal, single-core context — no concurrent access.
unsafe impl Send for Allocator {}

impl Allocator {
    /// Create a new (uninitialised) allocator.
    pub const fn new() -> Self {
        Self {
            fl_bitmap: 0,
            sl_bitmaps: [0; FL_INDEX_COUNT],
            blocks: [[ptr::null_mut(); SL_INDEX_COUNT]; FL_INDEX_COUNT],
            heap_size: 0,
            used: 0,
        }
    }

    /// Register a contiguous memory region as the heap.
    ///
    /// # Safety
    ///
    /// - `base` must be valid and writable for `size` bytes.
    /// - The region must be exclusively owned by the allocator.
    /// - Must not be called more than once.
    pub unsafe fn init(&mut self, base: *mut u8, size: usize) {
        if size < MIN_BLOCK_SIZE * 4 {
            return; // Too small to be useful.
        }

        self.heap_size = size;
        self.used = 0;

        // Place a sentinel block at the start.
        let block = base as *mut BlockHeader;
        let block_size = size - core::mem::size_of::<BlockHeader>();
        // Round down to alignment.
        let block_size = block_size & SIZE_MASK;

        unsafe {
            (*block).size_and_flags = block_size | BLOCK_FREE;
            (*block).prev_phys = ptr::null_mut();
            (*block).next_free = ptr::null_mut();
            (*block).prev_free = ptr::null_mut();
        }

        self.insert_free_block(block);
    }

    /// Allocate `size` bytes with the given alignment.
    ///
    /// Returns a pointer to the allocated region, or null on failure.
    /// The returned pointer is guaranteed to be at least `align`-byte aligned.
    pub fn alloc(&mut self, size: usize, align: usize) -> *mut u8 {
        let adjusted = adjust_size(size, align);
        if adjusted == 0 {
            return ptr::null_mut();
        }

        let block = self.find_suitable_block(adjusted);
        if block.is_null() {
            return ptr::null_mut();
        }

        self.remove_free_block(block);
        self.split_and_use(block, adjusted);

        unsafe { (*block).payload() }
    }

    /// Free a previously allocated pointer.
    ///
    /// # Safety
    ///
    /// `ptr` must have been returned by a prior call to [`alloc`](Allocator::alloc)
    /// and must not have been freed already.
    pub unsafe fn free(&mut self, ptr: *mut u8) {
        if ptr.is_null() {
            return;
        }

        let block = payload_to_header(ptr);
        unsafe {
            self.used -= (*block).size();
            (*block).set_free();
        }

        // Coalesce with adjacent free blocks.
        let block = unsafe { self.merge_next(block) };
        let block = unsafe { self.merge_prev(block) };

        self.insert_free_block(block);
    }

    /// Total heap bytes managed.
    pub fn heap_size(&self) -> usize { self.heap_size }

    /// Bytes currently in use (including block headers).
    pub fn used(&self) -> usize { self.used }

    /// Bytes available for allocation.
    pub fn free_bytes(&self) -> usize { self.heap_size - self.used }

    // ---- Internal ----

    fn insert_free_block(&mut self, block: *mut BlockHeader) {
        let size = unsafe { (*block).size() };
        let (fl, sl) = mapping(size);

        let head = self.blocks[fl][sl];
        unsafe {
            (*block).next_free = head;
            (*block).prev_free = ptr::null_mut();
            if !head.is_null() {
                (*head).prev_free = block;
            }
        }
        self.blocks[fl][sl] = block;
        self.fl_bitmap |= 1 << fl;
        self.sl_bitmaps[fl] |= 1 << sl;
    }

    fn remove_free_block(&mut self, block: *mut BlockHeader) {
        let size = unsafe { (*block).size() };
        let (fl, sl) = mapping(size);

        let next = unsafe { (*block).next_free };
        let prev = unsafe { (*block).prev_free };

        if !next.is_null() {
            unsafe { (*next).prev_free = prev };
        }
        if !prev.is_null() {
            unsafe { (*prev).next_free = next };
        } else {
            // Block was the head of this list.
            self.blocks[fl][sl] = next;
            if next.is_null() {
                self.sl_bitmaps[fl] &= !(1 << sl);
                if self.sl_bitmaps[fl] == 0 {
                    self.fl_bitmap &= !(1 << fl);
                }
            }
        }
    }

    fn find_suitable_block(&self, size: usize) -> *mut BlockHeader {
        let (mut fl, mut sl) = mapping(size);

        // Search the second-level bitmap for a block in this FL class.
        let sl_map = self.sl_bitmaps[fl] & (!0u16 << sl);
        if sl_map != 0 {
            sl = sl_map.trailing_zeros() as usize;
            return self.blocks[fl][sl];
        }

        // No block in this FL class — search higher FL classes.
        let fl_map = self.fl_bitmap & (!0u32 << (fl + 1));
        if fl_map == 0 {
            return ptr::null_mut(); // Out of memory.
        }
        fl = fl_map.trailing_zeros() as usize;
        sl = self.sl_bitmaps[fl].trailing_zeros() as usize;
        self.blocks[fl][sl]
    }

    fn split_and_use(&mut self, block: *mut BlockHeader, size: usize) {
        let block_size = unsafe { (*block).size() };
        let remaining = block_size - size;

        unsafe { (*block).set_used() };
        unsafe { (*block).set_size(size) };
        self.used += size;

        if remaining >= MIN_BLOCK_SIZE {
            let rest = unsafe { (*block).next_phys() };
            unsafe {
                (*rest).size_and_flags = remaining | BLOCK_FREE;
                (*rest).prev_phys = block;
            }
            self.insert_free_block(rest);
        }
    }

    unsafe fn merge_next(&mut self, block: *mut BlockHeader) -> *mut BlockHeader {
        let next = unsafe { (*block).next_phys() };
        if !next.is_null() && unsafe { (*next).is_free() } {
            self.remove_free_block(next);
            let merged_size = unsafe { (*block).size() + (*next).size() };
            unsafe { (*block).set_size(merged_size) };
        }
        block
    }

    unsafe fn merge_prev(&mut self, block: *mut BlockHeader) -> *mut BlockHeader {
        if unsafe { (*block).is_prev_free() } {
            let prev = unsafe { (*block).prev_phys };
            if !prev.is_null() {
                self.remove_free_block(prev);
                let merged_size = unsafe { (*prev).size() + (*block).size() };
                unsafe { (*prev).set_size(merged_size) };
                return prev;
            }
        }
        block
    }
}

// ============================================================================
// Helpers
// ============================================================================

/// Map a block size to first-level and second-level indices.
fn mapping(size: usize) -> (usize, usize) {
    if size < MIN_BLOCK_SIZE {
        return (0, 0);
    }
    let fl = (usize::BITS - 1 - size.leading_zeros()) as usize;
    let fl = fl.saturating_sub(FL_INDEX_SHIFT);
    let sl = (size >> (fl + FL_INDEX_SHIFT - SL_INDEX_COUNT_LOG2)) & (SL_INDEX_COUNT - 1);
    (fl.min(FL_INDEX_COUNT - 1), sl)
}

/// Round `size` up to include the block header and satisfy alignment.
fn adjust_size(size: usize, align: usize) -> usize {
    let size = size.max(MIN_BLOCK_SIZE);
    let align = align.max(core::mem::align_of::<BlockHeader>());
    // Add header overhead.
    let total = size + core::mem::size_of::<usize>() * 2;
    // Align up.
    (total + align - 1) & !(align - 1)
}

/// Convert a user-data pointer back to its block header.
fn payload_to_header(ptr: *mut u8) -> *mut BlockHeader {
    unsafe { ptr.sub(core::mem::size_of::<usize>() * 2) as *mut BlockHeader }
}
